Check for null when comparing using the most precise type.

spudstuff opened the pull request:
This allows null value parameters to be used in expressions such as `'a string' == null` where null is defined as `e.Parameters["null"] = null;`.

Some of my custom functions can return null and there are cases where I need to detect that. Currently the code is calling `GetType()` on the null and getting a null reference exception.

spudstuff committed:
Check for null when comparing using the most precise type. This allows null value parameters to be used in expressions such as "'a string' == null" where null is defined as e.Parameters["null"] = null.

spudstuff commented:
Hi @pitermarx - any issues with this PR? Love to get it merged and nuget updated.

pitermarx commented:
Thanks for the contribution. I'm returning from vacation today so I'll only check this in a few days, as I have some "work" stuff to get done right away.

Some quick questions: 

Was null not allowed as a variable name before or just the null value not allowed?
What was the behaviour before on the tests you added?
As you can see in the README, i'm commited to have no breaking changes on this.
From a quick read, is the change the same as doing only?
`Type mpt = GetMostPreciseType(a?.GetType(), b?.GetType());`
What version number should be bumped with this change?

spudstuff commented:
Hi,

> Was null not allowed as a variable name

No, null isn't a built-in parameter name:

```
System.ArgumentException : Parameter was not defined
Parameter name: null
```

But is easily defined yourself:

```
e.Parameters["null"] = null;
```

> or just the null value not allowed

I don't believe any of the built-in functions can return null. They're all math functions with the exception of IF and IN.

> What was the behaviour before on the tests you added?

Using this test:

```
var e = new Expression("null == 'a string'");
e.Parameters["null"] = null;
e.Evaluate();
```

Results in this exception:

```
System.NullReferenceException : Object reference not set to an instance of an object.
   at NCalc.Domain.EvaluationVisitor.CompareUsingMostPreciseType(Object a, Object b) in D:\NCalc-Edge\Evaluant.Calculator\Domain\EvaluationVisitor.cs:line 52
   at NCalc.Domain.EvaluationVisitor.Visit(BinaryExpression expression) in D:\NCalc-Edge\Evaluant.Calculator\Domain\EvaluationVisitor.cs:line 123
   at NCalc.Domain.BinaryExpression.Accept(LogicalExpressionVisitor visitor) in D:\NCalc-Edge\Evaluant.Calculator\Domain\BinaryExpression.cs:line 20
   at NCalc.Expression.Evaluate() in D:\NCalc-Edge\Evaluant.Calculator\Expression.cs:line 255
```

> From a quick read, is the change the same as doing only?
> Type mpt = GetMostPreciseType(a?.GetType(), b?.GetType());

Not quite, since that forces GetMostPreciseType to return `null` (it returns `a`, null in this case):

```
private static Type GetMostPreciseType(Type a, Type b)
{
    foreach (Type t in CommonTypes)
    {
        if (a == t || b == t)
        {
            return t;
        }
    }
    return a;
}
```

And `Convert.ChangeType(Object, Type)` does not accept null for the `Type` parameter:

```
System.ArgumentNullException : Value cannot be null.
Parameter name: conversionType
   at System.Convert.ChangeType(Object value, Type conversionType, IFormatProvider provider)
   at System.Convert.ChangeType(Object value, Type conversionType)
   at NCalc.Domain.EvaluationVisitor.CompareUsingMostPreciseType(Object a, Object b) in D:\TeamGit\NCalc-Edge\Evaluant.Calculator\Domain\EvaluationVisitor.cs:line 53
   at NCalc.Domain.EvaluationVisitor.Visit(BinaryExpression expression) in D:\TeamGit\NCalc-Edge\Evaluant.Calculator\Domain\EvaluationVisitor.cs:line 123
   at NCalc.Domain.BinaryExpression.Accept(LogicalExpressionVisitor visitor) in D:\TeamGit\NCalc-Edge\Evaluant.Calculator\Domain\BinaryExpression.cs:line 20
   at NCalc.Expression.Evaluate() in D:\TeamGit\NCalc-Edge\Evaluant.Calculator\Expression.cs:line 255
```

> As you can see in the README, i'm commited to have no breaking changes on this.

I can't be 100% on no breaking changes, but I can't see how anyone else might be using a `null` parameter, or they'd be seeing this problem too. Certainly none of the existing tests break with this change.

But we could always make it an option, I'd be happy with that. Defining the option can implement the new `CompareUsingMostPreciseType` behavior and maybe define a default parameter named `null` too.

```
Expression e = new Expression("'string' == null", EvaluateOptions.AllowNullParameter)
e.Evaluate();
```

> What version number should be bumped with this change?

1.4.5 if you're happy enough that you don't think anyone will be affected by the change for `null` variables. Maybe 1.5.0 if you want to add the option to set the behaviour (I can give that a shot if you like).

Cheers, Adam.

spudstuff committed:
Added AllowNullParameter option to define a null parameter and allow comparison of values to null. Set as an option to not affecting existing use of NCalc.

spudstuff commented:
Hi @pitermarx,

I added another commit to add the `AllowNullParameter` option, and some additional tests to confirm existing behaviour (throws `NullReferenceException`) when the option isn't set.

```
var e = new Expression("'a string' == null", EvaluateOptions.AllowNullParameter);
Assert.AreEqual(false, e.Evaluate());
```

Please just check I added the null parameter into the dictionary in the correct spot (`Expression.Evaluate()`).

Otherwise I think it's a good compromise - people can turn on support for null comparisons and the null parameter if they wish, and existing code isn't affected.

Cheers, Adam.

spudstuff committed:
Added some additional null checks from https://github.com/pitermarx/NCalc-Edge/pull/3

pitermarx commented:
I think this is definitely better. Keeps existing behaviour and adds a small feature. 
Let me think a bit about this but it looks good!

pitermarx commented on Evaluant.Calculator/Expression.cs:
```diff
@@ -196,6 +196,13 @@ public object Evaluate()
             visitor.EvaluateParameter += EvaluateParameter;
             visitor.Parameters = Parameters;
 
+            // Add a "null" parameter which returns null if configured to do so
```
nice touch

pitermarx commented on Evaluant.Calculator/Domain/EvaluationVisitor.cs:
```diff
@@ -29,7 +29,26 @@ public override void Visit(LogicalExpression expression)
+        /// <param name="a">Object a.</param>
+        /// <param name="b">Object b.</param>
+        /// <returns></returns>
+        private static Type GetMostPreciseType(object a, object b)
```
I'ts itching me to have .GetType inside a loop. It's not a big deal, it's just my OCD
Does this work?

```        private static Type GetMostPreciseType(Type a, Type b)
        {
            foreach (Type t in CommonTypes)
            {
                if (a == t || b == t)
                {
                    return t;
                }
            }

            return a ?? b;
        }

        public int CompareUsingMostPreciseType(object a, object b)
        {
            Type mpt = options.AllowNullParameter() 
                     // Allow nulls to be compared with other values
                     ? GetMostPreciseType(a?.GetType(), b?.GetType()) ?? typeof(object)
                     // No breaking changes fallback
                     : GetMostPreciseType(a.GetType(), b.GetType());
            return Comparer.Default.Compare(Convert.ChangeType(a, mpt), Convert.ChangeType(b, mpt));
        }
```

spudstuff replied:
Yep, that works fine. PR updated.

pitermarx replied:
can the overload GetMostPreciseType(object a, object b) be deleted?

spudstuff committed:
Remove .GetType() call from loop, fix typo.

spudstuff committed:
Removed unused overload.

spudstuff commented:
Sorry, now removed.

pitermarx commented:
merged. Thanks for the contribution

pitermarx commented:
https://www.nuget.org/packages/NCalc-Edge/1.5.0

spudstuff commented:
Love it, thanks mate!
